### PR TITLE
[FLINK-29472] Add first version of release scripts

### PR DESCRIPTION
**Pull request by zentol**
This PR adds the first version of shared release utilities for connectors.

The expectation is that the `release_utils` branch will be mounted as a git submodule into each connector repo.

See the README for the purpose of each script.

### CODE REVIEW

**Comment by zentol** [comment]
Note that we could just do a review round, and then try them out for the ES connector release before merging them.

File: check_environment.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+echo "Checking program availability:"
+check_program_available git
+check_program_available tar
+check_program_available rsync
+check_program_available gpg
+check_program_available perl
+check_program_available sed
+check_program_available svn
+check_program_available ${MVN}
```
Comment (MartijnVisser) [inline check_environment.sh:43]: Just to double check, any version of Maven for connectors should suffice, right? Doesn't need to be 3.2.5 like for Flink itself

File: README.md [hunk 2]
```diff
@@ -1 +1,79 @@
...
+for `@PublicEvolving` when used on snapshot branches an clears the list of exclusions.  
+This should be called after a release on the associated snapshot branch. If it was a minor release it should
+additionally be called on the `main` branch.
+
+# Common workflow
+
+1. run `publish_snapshot_branch.sh`
+2. do some development work on the created snapshot branch
+3. checkout a specific commit to create a release from
+4. run `check_environment.sh`
```
Comment (MartijnVisser) [inline README.md:73]: I think it makes more sense to have this as the first step and not the 4th. The `publish_snapshots_branch.sh` uses Maven and git, which are checked for availability in `check_environment.sh`.

File: README.md [hunk 3]
```diff
@@ -1 +1,79 @@
...
+## check_environment.sh
+
+Runs some pre-release checks for the current environment, for example that all required programs are available.  
+This should be run once at the start of the release process.
+
+## publish_snapshot_branch.sh
+
+Creates (and pushes!) a new snapshot branch for the current commit.  
+The branch name is automatically determined from the version in the pom.  
+This script should be called when work on a new major/minor version has started.
```
Comment (MartijnVisser) [inline README.md]:
```suggestion
This script should be called when work on a new major/minor version of the connector has started.
```

File: publish_git_tag.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+function create_release_tag {
+  cd "${SOURCE_DIR}"
+
+  version=$(get_pom_version)
+
+  tag=v${version}
+  if [ "$RC_NUM" != "none" ]; then
+    tag=${tag}-rc${RC_NUM}
+  fi
```
Comment (MartijnVisser) [inline publish_git_tag.sh]: Does it make sense to add a check here that verifies if the version doesn't contain SNAPSHOT anymore?

File: _init.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+set -o nounset
+set -o pipefail
+
+export SHELLOPTS
+
+###########################
+
+MVN=${MVN:-mvn}
+
+if [ "$(uname)" == "Darwin" ]; then
```
Comment (MartijnVisser) [inline _init.sh:33]: While I have Darwin, I can also both run `sha512sum` and `shasum -a 512`. Don't think that's an issue for this script's purpose though

**Review by MartijnVisser — APPROVED** [review]
This looks good. I think it's best to indeed test this with the Elasticsearch repo to see if everything works.

File: check_environment.sh (hunk 1 above)
Comment (zentol) [reply check_environment.sh:43]: This script doesn't neither enforces a maven version nor uses fancy features that would implicitly require a newer maven version.
This is all left to the connector.

File: _init.sh (hunk 5 above)
Comment (zentol) [reply _init.sh:33]: Doesn't hurt to have it I guess :shrug:
I suppose we just can't be sure this is the case in general.

File: README.md (hunk 2 above)
Comment (zentol) [reply README.md:73]: Basically the issue is that creating a snapshot branch will sometimes not be done by a release manager.
For example, after the ES 3.0 release we have a v3.0 branch and a main branch on 4.0-SNAPSHOT. v3.1 would be created by _someone_ when we want to make some change that requires a new minor version.

So all the other things the check does just aren't required.

File: publish_git_tag.sh (hunk 4 above)
Comment (zentol) [reply publish_git_tag.sh]: We could do that, yes.

File: _utils.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  clean_dir="${release_dir}/tmp-clean-clone"
+  # create a temporary git clone to ensure that we have a pristine source release
+  git clone "${source_dir}" "${clone_dir}"
+
+  rsync -a \
+    --exclude ".git" --exclude ".gitignore" --exclude ".gitattributes" --exclude ".gitmodules" --exclude ".github" \
+    --exclude ".idea" --exclude "*.iml" \
+    --exclude ".DS_Store" \
+    --exclude ".asf.yaml" \
+    --exclude "target" --exclude "tools/releasing/shared" \
```
Comment (zentol) [inline _utils.sh:52]: ideally the second exclusion would be more dynamic and figure out on its own what the `shared` directory is actually called.

File: README.md [hunk 7]
```diff
@@ -1 +1,79 @@
-This repository contains utilities for [Apache Flink](https://flink.apache.org/) connectors.
\ No newline at end of file
+This is a collection of release utils for [Apache Flink](https://flink.apache.org/) connectors.
+
+# Integration
+
+The scripts assume that they are integrated into a connector repo as a submodule into the connector repo
+under `tools/releasing/<some directory>`.
```
Comment (zentol) [inline README.md]:
```suggestion
under `tools/releasing/shared`.
```

File: stage_jars.sh [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+  echo "Deploying jars v${version} to repository.apache.org"
+  echo "To revert this step, login to 'https://repository.apache.org' -> 'Staging repositories' -> Select repository -> 'Drop'"
+
+  clone_dir=$(create_pristine_source "${SOURCE_DIR}" "${RELEASE_DIR}")
+  cd "${clone_dir}"
+  set_pom_version "${version}"
+
+  options="-Prelease,docs-and-source -DskipTests -DretryFailedDeploymentCount=10"
+  ${MVN} clean deploy ${options}
```
Comment (zentol) [inline stage_jars.sh]: This should set the `flink.version` property so we actually compile against the correct Flink version.
May require the user to provide a full Flink version from which we extract the minor version.

**Comment by leonardBang** [comment]
Thanks @zentol and @MartijnVisser  for driving this, but I saw `dev` mail list received all PR updates information from this repo, is there some settings incorrect?

**Comment by MartijnVisser** [comment]
> Thanks @zentol and @MartijnVisser for driving this, but I saw `dev` mail list received all PR updates information from this repo, is there some settings incorrect?

Yes, we need to add an `.asf.yaml` file to this repo. Will do in a moment :)

**Comment by leonardBang** [comment]
Happy to see you start working @zentol , Could you take a look my comments here? https://github.com/apache/flink/pull/21227 sorry to reply under this issue but I've tried ping you on slack but no response.

File: README.md [hunk 9]
```diff
@@ -1 +1,79 @@
...
+12. run `update_japicmp_configuration.sh` (on snapshot branch, and maybe `main`)
+
+# Script naming conventions
+
+| Prefix  | Meaning                                                                |
+|---------|------------------------------------------------------------------------|
+| check   | Verifies conditions without making any changes.                        |
+| update  | Applies modifications locally to the current branch.                   |
+| stage   | Publishes an artifact to an intermediate location for voting purposes. |
+| publish | Releases an artifact to a user-facing location.                        |
```
Comment (zentol) [inline README.md]:
```suggestion
| release | Publishes an artifact to a user-facing location.                        |
```
It's a bit of a mess with "publishing" having 2 different meanings.

**Review by MartijnVisser — APPROVED** [review]

**Pull request merged**